src/providers/php.rs

```diff
@@ -0,0 +1,211 @@
+use std::collections::HashMap;
+
+use serde::{Deserialize, Serialize};
+
+use crate::nixpacks::{
+    app::App,
+    environment::EnvironmentVariables,
+    nix::pkg::Pkg,
+    phase::{BuildPhase, InstallPhase, SetupPhase, StartPhase},
+};
+
+use super::{node::NodeProvider, Provider};
+
+pub struct PhpProvider;
+
+impl Provider for PhpProvider {
+    fn name(&self) -> &str {
+        "php"
+    }
+
+    fn detect(
+        &self,
+        app: &crate::nixpacks::app::App,
+        _env: &crate::nixpacks::environment::Environment,
+    ) -> anyhow::Result<bool> {
+        Ok(app.includes_file("composer.json") || app.includes_file("index.php"))
+    }
+
+    fn setup(
+        &self,
+        app: &crate::nixpacks::app::App,
+        env: &crate::nixpacks::environment::Environment,
+    ) -> anyhow::Result<Option<crate::nixpacks::phase::SetupPhase>> {
+        let nodejs = NodeProvider {};
+
+        let php_pkg = match self.get_php_package(app) {
+            Ok(Some(php_package)) => php_package,
+            _ => "php".to_string(),
+        };
+        let mut pkgs = vec![
+            Pkg::new(&php_pkg),
+            Pkg::new("perl"),
+            Pkg::new("nginx"),
+            Pkg::new(&format!("{}Packages.composer", &php_pkg)),
+        ];
+        if let Ok(php_extensions) = self.get_php_extensions(app) {
+            for extension in php_extensions {
+                pkgs.push(Pkg::new(&format!("{}Extensions.{}", &php_pkg, extension)));
+            }
+        }
+
+        if let Ok(true) = nodejs.detect(app, env) {
+            if let Ok(Some(mut node_setup)) = nodejs.setup(app, env) {
+                pkgs.append(&mut node_setup.pkgs);
+            }
+        }
+
+        Ok(Some(SetupPhase::new(pkgs)))
+    }
+
+    fn install(
+        &self,
+        app: &crate::nixpacks::app::App,
+        env: &crate::nixpacks::environment::Environment,
+    ) -> anyhow::Result<Option<crate::nixpacks::phase::InstallPhase>> {
+        let mut cmd = String::new();
+        cmd.push_str("mkdir -p /var/log/nginx && mkdir -p /var/cache/nginx");
+        let nodejs = NodeProvider {};
```

Comment:
This is really interesting and a good forcing function to allow matching multiple providers at once. This is okay for now but I think allowing more than one provider would clean this up quite a bit.